File: 49_Group_anagrams.py
# https://leetcode.com/problems/group-anagrams/
# 49. Group Anagrams

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram_groups = defaultdict(list)

        for s in strs:
            sorted_s = ''.join(sorted(s))
            anagram_groups[sorted_s].append(s)

        return list(anagram_groups.values())

File: 49_Group_anagrams.cpp
// https://leetcode.com/problems/group-anagrams/
// 49. Group Anagrams

class Solution
{
public:
    std::vector<std::vector<std::string>> groupAnagrams(std::vector<std::string> &strs)
    {
        std::unordered_map<std::string, std::vector<std::string>> anagramGroups;

        for (const std::string &s : strs)
        {
            std::string sorted_s = s;
            std::sort(sorted_s.begin(), sorted_s.end());
            anagramGroups[sorted_s].push_back(s);
        }

        std::vector<std::vector<std::string>> result;
        for (const auto &entry : anagramGroups)
        {
            result.push_back(entry.second);
        }

        return result;
    }
};